Setup project infrastructure (#2)
filecoin-project/go-ds-versioning: README.md

# go-ds-versioning
[![](https://img.shields.io/badge/made%20by-Protocol%20Labs-blue.svg?style=flat-square)](http://ipn.io)
[![CircleCI](https://circleci.com/gh/filecoin-project/go-ds-versioning.svg?style=svg)](https://circleci.com/gh/filecoin-project/go-ds-versioning)
[![codecov](https://codecov.io/gh/filecoin-project/go-ds-versioning/branch/master/graph/badge.svg)](https://codecov.io/gh/filecoin-project/go-ds-versioning)

Tools for building migratable datastore

## Description

This module provides building blocks for migrating key-value datastores that match the interfaces defined in [go-datastore](https://github.com/ipfs/go-datastore)

## Table of Contents
* [Background](#background)
* [Installation](#installation)
* [Usage](#usage)
    * [Defining migrations](#defining-migrations)
    * [Executing migrations on Datastores, StateStores, and Finite State Machine groups](#executing-migrations)
* [Architecture](#architecture)
* [Contribute](#contribute)

## Background

In the Filecoin ecosystem, we maintain a number of key-value datastores that

> track the off-chain states fo various components of Filecoin. As the protocol evolves, we will need to migrate these datastores. The Filecoin protocol requires nodes on the network maintain near constant uptime in order to meet proving requirements. We'd like to be able to migrate off-chain states while taking down only the components of the node that use these states.

Reviewer: ```suggestion
track the off-chain states of various components of Filecoin. As the protocol evolves, we will need to migrate these datastores. The Filecoin protocol requires nodes on the network maintain near constant uptime in order to meet proving requirements. We'd like to be able to migrate off-chain states while taking down only the components of the node that use these states.
```

This library provides two core components:

1. Tools for defining migrations. `go-datastore` is a freeform key-value database, so a migration could potentially mean any kind of transformation on
the database. However, most commonly, we want to migrate one record type to another. This library provides tools for doing that easily, and for composing these migrations together with versions to build a system for a versioned datastore that can be migrated easily with minimal potential for data-loss

2. Tools for migrating datastores, statestores, and finite state machines transparently, without changing the interfaces for those stores. We define versions of these stores that simply error until they are migrated. This means we can run migrations seperately and when they finish (if they finish successfully), the store comes online.

## Installation

**Requires go 1.14**

> Install the module in your package or app with `go get "github.com/filecoin-project/go-ds-versioning/datatransfer"`

Reviewer: ```suggestion
Install the module in your package or app with `go get "github.com/filecoin-project/go-ds-versioning"`
```

```

There's a lot happening there so let's take a look at this. We build a migration here:

```golang
builder := versioned.NewVersionedBuilder(MigrateFruitBasket, versioning.VersionKey("1")),
```

We're defining a migration that will use MigrateFruitBasket as its transformation function and will migration to version `1`. While not shown here, there are other things we can also do with this builder:

```golang
// if we want to provide a mechanism to reverse this migration
builder := builder.Reversable(UnMigrateFruitBasket)

// after our initial version, we always want to make sure we define the old version we're migrating from
version2Builder := version2Builder.OldVersion("1")

// we may need to for whatever reason, leave certain keys out of a migration
// (i.e. delete them, or we may need to exclude some keys that are at the same
// namespace hierarchy in our initial migration)
builder := builder.FilterKeys("rotten-fruit-basket")
```

We're assuming we'll probably define all our migrations in one place, so we make a `BuilderList` -- a list of migration definitions assembled using our builder interface. Typically you can just put your builders inline in the BuilderList.

> Finally, to turn these into actual migrations, we call `.Build()` on the `BuilderList` -- this will that our migrations are valid. In particular, every migration function must have the form:

Reviewer: ```suggestion
Finally, to turn these into actual migrations, we call `.Build()` on the `BuilderList` -- this will ensure that our migrations are valid. In particular, every migration function must have the form:
```

```golang
func <T extends cbg.CBORUnmarshaller, U extends cbg.CBORMarshaller>(old T) (new U, error)
```

### Executing Migrations

Let's say we are using a [go-statestore](https://github.com/filecoin-project/go-statestore) for our fruit baskets:

```golang
import (
    "github.com/ipfs/go-datastore"
    "github.com/filecoin-project/go-statestore"
)

var ds datastore // this was setup elsewhere
fruitBaskets := statestore.New(ds)
```

To make it migratable, we replace it with a versioned statestore, with our migrations:

```golang
import (
    "github.com/ipfs/go-datastore"
    versioning "github.com/filecoin-project/go-ds-versioning/pkg"
    "github.com/filecoin-project/go-ds-versioning/pkg/statestore"
)

var ds datastore // this was setup elsewhere
var migrations versioning.VersionedMigrationList // the migrations we setup previously

fruitBaskets, migrateFruitBaskets := statestore.NewVersionedStateStore(ds, migrations, versioning.VersionKey("1"))
```

> Here we're setting up a migrated statestore using the datastore, a list of migrations, and a target version. (target cause in certain cases we want to support migrated DOWN).

Reviewer: ```suggestion
Here we're setting up a migrated statestore using the datastore, a list of migrations, and a target version. (target cause in certain cases we want to migrate to a particular version, either UP or DOWN).
```
